working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,559 @@
+
+      # Assert
+      assert response.status_code == 400
+      assert response_body == {"message":"book cat invalid"}
+
+  def test_get_one_book(client, two_saved_books):
+      # Act
+      response = client.get("/books/1")
+      response_body = response.get_json()
+
+      # Assert
+      assert response.status_code == 200
+      assert response_body == {
+          "id": 1,
+          "title": "Ocean Book",
+          "description": "watr 4evr"
+      }
+
+  def test_create_one_book(client):
+      # Act
+      response = client.post("/books", json={
+          "title": "New Book",
+          "description": "The Best!"
+      })
+      response_body = response.get_json()
+
+      # Assert
+      assert response.status_code == 201
+      assert response_body == "Book New Book successfully created"
+
+  def test_create_one_book_no_title(client):
+      # Arrange
+      test_data = {"description": "The Best!"}
+
+      # Act & Assert
+      with pytest.raises(KeyError, match='title'):
+          response = client.post("/books", json=test_data)
+
+  def test_create_one_book_no_description(client):
+      # Arrange
+      test_data = {"title": "New Book"}
+
+      # Act & Assert
+      with pytest.raises(KeyError, match = 'description'):
+          response = client.post("/books", json=test_data)
+
+  def test_create_one_book_with_extra_keys(client, two_saved_books):
+      # Arrange
+      test_data = {
+          "extra": "some stuff",
+          "title": "New Book",
+          "description": "The Best!",
+          "another": "last value"
+      }
+
+      # Act
+      response = client.post("/books", json=test_data)
+      response_body = response.get_json()
+
+      # Assert
+      assert response.status_code == 201
+      assert response_body == "Book New Book successfully created"
+
+  def test_validate_book_with_valid_id(two_saved_books):
+      # Arrange
+      book_id = 1
+
+      # Act
+      result = validate_book(book_id)
+
+      # Assert
+      assert result.id == book_id
+
+  def test_validate_book_with_invalid_id():
+      # Arrange
+      book_id = "cat"
+
+      # Act & Assert
+      with pytest.raises(Exception) as e:
+          validate_book(book_id)
+          assert str(e.value) == "book cat invalid"
+
+  def test_validate_book_with_nonexistent_id():
+      # Arrange
+      book_id = 3
+
+      # Act & Assert
+      with pytest.raises(Exception) as e:
+          validate_book(book_id)
+          assert str(e.value) == "book 3 not found"
+  ```
+</details>
+
+#### Testing `update_book`
+
+With `validate_book` we were lucky and Copilot outlined all of the scenarios that we intended to test, but that will not always be the case. We won't go step by step again, but we want to take the same initial approach to testing our remaining dependencies `update_book` & `delete_book`. 
+
+Let's use the Copilot inline chat to ask for help generating tests for the `update_book` function and take a look at what Copilot suggested below:
```

Comment:
- since I just used "let's" above, rephrased to avoid sounding repetative
```suggestion
We'll start by using Copilot inline chat to ask for help generating tests for the `update_book` function. Here's what Copilot suggested:
```